### PR TITLE
EPS-11: Implement search and sorting logic

### PR DESCRIPTION
**Pull request by vladqf**
[JIRA-11](https://ifortex-team-java1.atlassian.net/browse/EPS-11)
- Added search and sorting capability,
- Fixed issue with passing non-existing student IDs and name validation during course create/update.

### CODE REVIEW

File: src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java [hunk 1]
```diff
@@ -104,6 +136,16 @@ public void update(long courseId, Map<CourseField, Object> fieldMap) {
     jdbcTemplate.update(sqlBuilder.toString(), values.toArray());
   }
 
+  @Override
+  public List<Student> getExistingStudents(List<Long> studentIds) {
+    if (studentIds.isEmpty()) {
+      return List.of();
+    }
+    String sql = "SELECT * FROM students WHERE id IN (:studentIds)";
```
Comment (PavelShynkarou) [inline src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:144]: As far as I can see the main reason to provide new Jdbc bean is list ids usaging? But from my perspective it still can be implemented with jdbcTemplate itself. Lets support one style of querying.

File: src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java [hunk 2]
```diff
@@ -48,6 +57,41 @@ public List<Course> findAll() {
...
+    }
+
+    boolean hasSort = false;
+    if (dto.getSortByDate() != null) {
+      sqlBuilder.append(" ORDER BY c.start_date ").append(dto.getSortByDate().name().toUpperCase());
+      hasSort = true;
+    }
+
+    if (dto.getSortByName() != null) {
+      if (hasSort) {
```
Comment (PavelShynkarou) [inline src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:85]: ` sqlBuilder.append(hasSort ? ", " : " ORDER BY ")
              .append("c.name ")
              .append(dto.getSortByName().name().toUpperCase());`

**Review by PavelShynkarou — CHANGES_REQUESTED** [review]

File: src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java (hunk 1 above)
Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:144]: This is how it'll look like:
StringBuilder sqlBuilder = new StringBuilder("SELECT * FROM students WHERE id IN (");
    sqlBuilder.append(String.join(", ", studentIds.stream().map(Object::toString).toList()));
    sqlBuilder.append(")");
    
Not sure, which approach is better

File: src/main/java/com/ifortex/internship/dto/FilterSortDto.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.ifortex.internship.dto;
+
+
```
[the remark this answers is not in the document]
Comment (vladqf) [reply src/main/java/com/ifortex/internship/dto/FilterSortDto.java]: Fixed

File: src/main/java/com/ifortex/internship/repository/CourseRepository.java [hunk 4]
```diff
@@ -26,6 +27,14 @@ public interface CourseRepository {
    */
   List<Course> findAll();
 
+  /**
+   * Retrieves courses based on applied filters and sorting criteria.
```
[the remark this answers is not in the document]
Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/CourseRepository.java]: fixed

File: src/main/java/com/ifortex/internship/repository/CourseRepository.java [hunk 5]
```diff
@@ -48,6 +57,15 @@ public interface CourseRepository {
    */
   void update(long courseId, Map<CourseField, Object> fields);
 
+  /**
+   * Updates the students associated with a specific course based on new and existing student IDs.
```
[the remark this answers is not in the document]
Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/CourseRepository.java]: Done

File: src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java (hunk 2 above)
Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java]: Done

File: src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java (hunk 1 above)
Comment (PavelShynkarou) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:140]: What if something in this way ?
    String placeholders = studentIds.stream()
                                    .map(id -> "?")
                                    .collect(Collectors.joining(",", "(", ")"));

    String sql = "SELECT * FROM students WHERE id IN " + placeholders;
    return jdbcTemplate.query(sql, studentIds.toArray(), (rs, rowNum) -> 
        new Student(
            rs.getLong("id"),
            rs.getString("name")
        )
    );

Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:140]: Do we need to map id to "?" ? We can do smth like this:
image

Comment (PavelShynkarou) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:140]: You gonna substitute ids themself directly without prevent SQL injection and ignore SQL standarts ?

Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java:140]: Are there any injections possible? If so, your version will be better

Comment (vladqf) [reply src/main/java/com/ifortex/internship/repository/impl/JdbcCourseRepository.java]: Done

**Review by PavelShynkarou — APPROVED** [review]

**Review by AndreySemeynikovIF — APPROVED** [review]

**Pull request merged**